corehq/apps/es/fake/README.md

```diff
@@ -0,0 +1,73 @@
+(See http://stackoverflow.com/a/346440/240553 for the difference between fakes, mocks, and stubs.)
+
+`ESQueryFake` and its subclasses (`UserESFake`, etc.) do just this for the `ESQuery`
+classes. Whereas the real classes hand off the work to an Elasticsearch cluster,
+the fakes do the filtering, sorting, and slicing in python memory, which is lightweight
+and adequate for tests. Beware that this method is, of course,
+inadequate for assuring that the `ESQuery` classes themselves are producing
+the correct Elasticsearch queries, and also introduces the potential for bugs to go
+unnoticed because of bugs in `ESQueryFake` classes themselves. But assuming correct
+implementations of the fakes, it does an good job of testing the calling code,
+which is usually the primary subject of a test.
+
+The anatomy of a fake is something like this:
+- For each real `ESQuery` subclass (I'll use `UserES` as the example),
+  there is a corresponding fake (`UserESFake`).
+  In cases where such a fake does not exist when you need it,
+  follow the instructions below for getting started on a new fake.
+- For each filter method or public method used on the `ESQuery` base class
+  a method should exist on `ESQueryFake` that has the same behavior
+- For each filter method on `UserES`, a method should exist on `UserESFake`
+  that has the same behavior.
+
+New fakes and methods are implemented only as actually needed by tests
+(otherwise it's very difficult to be confident the implementations are correct),
+so until some mythical future day in which all code that relies on ES goes through
+an `ESQuery` subclass and is thoroughly tested, the fake implementations are
+intentionally incomplete. As such, an important part of their design is that they alert
+their caller (the person using them to write a test) if the code being tested calls a
+method on the fake that is not yet implemented. Since more often than not a number of
+methods will need to be added for the same test, the fakes currently are designed to have
+each call to an unimplemented filter result in a no-op, and will output a logging statement
+telling the caller how to add the missing method. This lets the caller run the test once
+to see a print out of every missing function, which they can then add in one go and re-run
+the tests. (The danger is that they will miss the logging output; however in cases where
+a filter method is missing, it is pretty likely that the test will fail which will prod
+them to look further and find the logging statement.)
+
+## How to set up your test to use ES fakes
+
+Patch your test to use `UserESFake`(assuming you want to patch `UserES`):
+
+```diff
++ @mock.patch('corehq.apps.es.UserES', UserESFake)
+  class MyTest(SimpleTestCase):
+      def setUp(self):
+...
++         UserESFake.save_doc(user._doc)
+...
+      def tearDown(self):
+          UserESFake.reset_docs()
+```
+
+Any file that uses your target `ESQuery` subclass must have all top-level imports of the
```

Comment:
Instead of this, in your call to `@mock.patch` you should patch the object where it is _looked up_ not where it is _defined_. i.e. you would batch the object in the place into which it has been imported. I'm not explaining myself properly, but here is an example.

Trying to mock `UserES` in `corehq.apps.analytics.tasks`.
```py
# in file: corehq/apps/analytics/tasks.py
from corehq.apps.es.users import UserES
# do a thing with UserES
```
For your `@mock` you would then do:
```py
@mock.patch('corehq.apps.analytics.tasks.UserES')
def test_thing(self, fake_user_es):
    self.assert_thing_is_a_dream(fake_user_es)
```
You can see more in my notes from my [Mock Brownbag](https://gist.github.com/proteusvacuum/ab6dbb95ed210fb55ada#file-mock_brownbag-py-L99) (which I'm going to transform into a blog post soonish)

Additionally, you can use a context manager with mock.
 ```py
def test_thing(self):
    with mock.patch('corehq.apps.analytics.tasks.UserES'):
         self.assert_thing_is_a_dream(UserES)
    self.assert_thing_is_real(UserES)
```